File: frontend/src/constants/registration.ts
// 报名状态枚举（与后端 backend/internal/constants/registration.go 保持一致）
export const RegistrationStatus = {
  REGISTERED: 'registered',
  CANCELLED: 'cancelled',
  CHECKED_IN: 'checked_in',
  WAITLISTED: 'waitlisted',
} as const

export const RegistrationStatusText: Record<string, string> = {
  [RegistrationStatus.REGISTERED]: '已报名',
  [RegistrationStatus.CANCELLED]: '已取消',
  [RegistrationStatus.CHECKED_IN]: '已签到',
  [RegistrationStatus.WAITLISTED]: '候补中',
}

export const ReviewStatus = {
  PENDING: 'pending',
  APPROVED: 'approved',
  REJECTED: 'rejected',
} as const

export const ReviewStatusText: Record<string, string> = {
  [ReviewStatus.PENDING]: '待审核',
  [ReviewStatus.APPROVED]: '已通过',
  [ReviewStatus.REJECTED]: '已拒绝',
}

export const CheckInMethodText: Record<string, string> = {
  voucher: '凭证签到',
  scan: '扫码签到',
}

// 综合展示：审核驳回会释放名额（status=cancelled + review_status=rejected），
// 名单中需显示为“已驳回”，与用户主动取消区分开。
export function displayRegistrationText(row: { status: string; review_status: string }): string {
  if (row.status === RegistrationStatus.CANCELLED && row.review_status === ReviewStatus.REJECTED) {
    return '已驳回'
  }
  return RegistrationStatusText[row.status] ?? row.status
}

export function displayRegistrationTag(row: { status: string; review_status: string }): string {
  if (row.status === RegistrationStatus.CANCELLED && row.review_status === ReviewStatus.REJECTED) {
    return 'danger'
  }
  if (row.status === RegistrationStatus.CHECKED_IN) return 'success'
  if (row.status === RegistrationStatus.CANCELLED) return 'info'
  if (row.status === RegistrationStatus.WAITLISTED) return 'warning'
  return 'primary'
}

// 可取消的报名：正式报名与候补中；已取消/已签到不可取消。
export function isCancellable(status: string): boolean {
  return status === RegistrationStatus.REGISTERED || status === RegistrationStatus.WAITLISTED
}

File: backend/internal/constants/registration.go
package constants

// RegistrationStatus 报名状态枚举。
const (
	RegistrationStatusRegistered = "registered"
	RegistrationStatusWaitlisted = "waitlisted"
	RegistrationStatusCancelled  = "cancelled"
	RegistrationStatusCheckedIn  = "checked_in"
)

// RegistrationStatusValues 全部报名状态值。
var RegistrationStatusValues = []string{
	RegistrationStatusRegistered, RegistrationStatusWaitlisted,
	RegistrationStatusCancelled, RegistrationStatusCheckedIn,
}

// OccupiedRegistrationStatuses 占用活动名额的报名状态（候补/已取消不占名额）。
var OccupiedRegistrationStatuses = []string{
	RegistrationStatusRegistered, RegistrationStatusCheckedIn,
}

// ReviewStatus 报名审核状态枚举。
const (
	ReviewStatusPending  = "pending"
	ReviewStatusApproved = "approved"
	ReviewStatusRejected = "rejected"
)

// ReviewStatusValues 全部审核状态值。
var ReviewStatusValues = []string{ReviewStatusPending, ReviewStatusApproved, ReviewStatusRejected}

// CheckInMethod 签到方式枚举。
const (
	CheckInMethodVoucher = "voucher"
	CheckInMethodScan    = "scan"
)

// IsValidRegistrationStatus 校验报名状态。
func IsValidRegistrationStatus(s string) bool {
	for _, v := range RegistrationStatusValues {
		if v == s {
			return true
		}
	}
	return false
}

// IsOccupiedRegistrationStatus 判断报名状态是否占用活动名额。
func IsOccupiedRegistrationStatus(s string) bool {
	for _, v := range OccupiedRegistrationStatuses {
		if v == s {
			return true
		}
	}
	return false
}

// IsValidReviewStatus 校验审核状态。
func IsValidReviewStatus(s string) bool {
	for _, v := range ReviewStatusValues {
		if v == s {
			return true
		}
	}
	return false
}
